namer/consul/src/main/scala/io/buoyant/namer/consul/LookupCache.scala

```diff
@@ -0,0 +1,79 @@
+package io.buoyant.namer.consul
+
+import com.twitter.finagle._
+import com.twitter.util._
+import io.buoyant.consul.v1
+
+/**
+ * A helper supporting service resolution in consul, caching results
+ * and watching for new updates.
+ */
+private[consul] class LookupCache(
+  consulApi: v1.ConsulApi,
+  agentApi: v1.AgentApi,
+  setHost: Boolean = false
+) {
+
+  def apply(
+    dcName: String,
+    svcKey: SvcKey,
+    id: Path,
+    residual: Path
+  ): Activity[NameTree[Name]] = {
+    log.debug("consul lookup: %s %s", id.show)
+    watchDc(dcName).map { services =>
+      services.get(svcKey) match {
+        case None =>
+          log.debug("consul dc %s service %s missing", dcName, svcKey)
+          NameTree.Neg
+
+        case Some(addr) =>
+          log.debug("consul ns %s service %s found + %s", dcName, svcKey, residual.show)
+          NameTree.Leaf(Name.Bound(addr, id, residual))
+      }
+    }
+  }
+
+  private[this] lazy val domain: Activity[Option[String]] =
```

Comment:
We'd just wrap that future in an Activity below to compose it in watchDc, so I don't know that it really makes a difference